tests/delayed_rejection_mh_producer_01.cc

```diff
@@ -0,0 +1,67 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2019 by the SampleFlow authors.
+//
+// This file is part of the SampleFlow library.
+//
+// The SampleFlow library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of SampleFlow.
+//
+// ---------------------------------------------------------------------
+
+
+// A simple test for the Delayed Rejection MH producer.
+// Copied from the metropolis_hastings_producer_01 test.
+
+
+#include <iostream>
+#include <sampleflow/producers/delayed_rejection_mh.h>
+#include <sampleflow/consumers/stream_output.h>
+#include <valarray>
+#include <random>
+#include <cmath>
+
+using SampleType = double;
+
+
+// Use a (non-normalized) probability distribution that increases left
+// to right.
+double log_likelihood (const SampleType &x)
+{
+  return x+1;
+}
+
+
+// Always move to the right when trying to find a new trial sample.
+std::pair<SampleType,double> perturb (const SampleType &x, const std::vector<SampleType> y)
+{
+  // Return both the new sample and the ratio of proposal distribution
+  // probabilities. We're moving the sample to the right, so that ratio
+  // is actually infinity, but we can lie about it for the purposes of
+  // this test.
+  return {x+1, 1.};
```

Comment:
I'm not sure this is right. The fact that the function lies about what the ratio of probabilities are, makes it probably so that it doesn't quite lead to correct results. It may be better to sample from something like a Gaussian, and compute the mean of that.

Reply:
Or you can leave the test as is and just go with it. Returning a ratio of 1 means that the delayed rejection simply always accepts the very first sample. That's not wrong, and in some sense just checks a specific part of the algorithm. So maybe it does make sense to just keep the test.